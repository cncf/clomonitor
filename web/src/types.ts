export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
