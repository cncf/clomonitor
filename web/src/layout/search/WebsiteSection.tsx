import classNames from 'classnames';
import { isUndefined } from 'lodash';
import { useEffect, useState } from 'react';
import { IoGlobeOutline } from 'react-icons/io5';

import { BaseRepository, Repository } from '../../types';
import DropdownOnHover from '../common/DropdownOnHover';
import ExternalLink from '../common/ExternalLink';
import styles from './WebsiteSection.module.css';

interface Props {
  repositories: BaseRepository[] | Repository[];
  onlyIcon?: boolean;
}

const WebsiteSection = (props: Props) => {
  const [websites, setWebsites] = useState<string[]>([]);
  const isOnlyIcon = !isUndefined(props.onlyIcon) && props.onlyIcon;

  useEffect(() => {
    let urls: string[] = [];
    props.repositories.forEach((repo: BaseRepository | Repository) => {
      if (repo.hasOwnProperty('report')) {
        const currentRepo = repo as Repository;
        if (
          currentRepo.report &&
          currentRepo.report.data &&
          currentRepo.report.data.documentation &&
          currentRepo.report.data.documentation.website &&
          currentRepo.report.data.documentation.website.url
        ) {
          urls.push(currentRepo.report.data.documentation.website.url);
        }
      } else if (!isUndefined(repo.website_url)) {
        urls.push(repo.website_url);
      }
    });
    setWebsites(urls);
  }, [props.repositories]); /* eslint-disable-line react-hooks/exhaustive-deps */

  if (websites.length === 0) return null;

  return (
    <div className="ms-3">
      {websites.length === 1 ? (
        <ExternalLink label="Website link" href={websites[0]}>
          <div
            className={classNames('d-flex flex-row align-items-center', styles.link, {
              [`text-muted ${styles.onlyIcon}`]: isOnlyIcon,
            })}
          >
            <IoGlobeOutline className={styles.icon} />
            {!isOnlyIcon && <div className="ms-1">Website</div>}
          </div>
        </ExternalLink>
      ) : (
        <DropdownOnHover
          width={250}
          linkContent={
            <div
              className={classNames('d-flex flex-row align-items-center', styles.link, {
                [`text-muted ${styles.onlyIcon}`]: isOnlyIcon,
              })}
            >
              <IoGlobeOutline className={styles.icon} />
              {!isOnlyIcon && <div className="ms-1">Websites</div>}
            </div>
          }
        >
          <>
            {websites.map((url: string, index: number) => {
              return (
                <div key={`repo_${index}`} className={`d-flex flex-row align-items-center my-1 ${styles.link}`}>
                  <IoGlobeOutline className={`me-2 position-relative ${styles.miniIcon}`} />
                  <div className="truncateWrapper">
                    <ExternalLink
                      label="Website link"
                      href={url}
                      className={`d-block text-truncate text-dark ${styles.link}`}
                      visibleExternalIcon
                    >
                      <div className="text-truncate">{url}</div>
                    </ExternalLink>
                  </div>
                </div>
              );
            })}
          </>
        </DropdownOnHover>
      )}
    </div>
  );
};

export default WebsiteSection;
