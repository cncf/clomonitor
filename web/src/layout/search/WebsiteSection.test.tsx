import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { CheckSet } from '../../types';
import WebsiteSection from './WebsiteSection';

const user = userEvent.setup({ delay: null });

const defaultPropsOneRepo = {
  repositories: [
    {
      check_sets: [CheckSet.Community, CheckSet.Code],
      name: 'artifact-hub',
      url: 'https://github.com/artifacthub/hub',
      website_url: 'https://artifacthub.io',
    },
  ],
};

const defaultPropsRepos = {
  repositories: [
    {
      check_sets: [CheckSet.Code],
      name: 'sdk-go',
      url: 'https://github.com/cloudevents/sdk-go',
    },
    {
      check_sets: [CheckSet.Code],
      name: 'sdk-javascript',
      url: 'https://github.com/cloudevents/sdk-javascript',
    },
    {
      check_sets: [CheckSet.Code],
      name: 'sdk-csharp',
      url: 'https://github.com/cloudevents/sdk-csharp',
    },
    {
      check_sets: [CheckSet.Code],
      name: 'sdk-python',
      url: 'https://github.com/cloudevents/sdk-python',
    },
    {
      check_sets: [CheckSet.Community],
      name: 'spec',
      url: 'https://github.com/cloudevents/spec',
      website_url: 'https://cloudevents.io',
    },
    {
      check_sets: [CheckSet.Code],
      name: 'sdk-java',
      url: 'https://github.com/cloudevents/sdk-java',
      website_url: 'https://cloudevents.io/test',
    },
  ],
};

describe('WebsiteSection', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<WebsiteSection {...defaultPropsRepos} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders with more than one website url', async () => {
      jest.useFakeTimers();

      render(<WebsiteSection {...defaultPropsRepos} />);

      const content = screen.getByText('Websites');
      expect(content).toBeInTheDocument();
      const dropdown = screen.getByRole('complementary');

      expect(dropdown).not.toHaveClass('show');

      await user.hover(content);

      act(() => {
        jest.advanceTimersByTime(100);
      });

      expect(dropdown).toHaveClass('show');
      expect(screen.getAllByRole('link', { name: 'Website link' })).toHaveLength(2);
    });

    it('renders with one url', () => {
      render(<WebsiteSection {...defaultPropsOneRepo} />);

      expect(screen.getByText('Website')).toBeInTheDocument();
      expect(screen.getByRole('link', { name: 'Website link' })).toBeInTheDocument();
      expect(screen.getByRole('link', { name: 'Website link' })).toHaveAttribute('href', 'https://artifacthub.io');
    });
  });
});
