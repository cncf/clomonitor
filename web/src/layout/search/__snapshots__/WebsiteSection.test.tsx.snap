// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebsiteSection creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="ms-3"
  >
    <div>
      <div
        class="position-absolute"
      >
        <div>
          <div
            class="dropdown-menu rounded-0 text-wrap dropdown"
            role="complementary"
            style="width: 250px;"
          >
            <div
              class="ps-3 pe-2 pt-1"
            >
              <div
                class="d-flex flex-row align-items-center my-1 link"
              >
                <svg
                  class="me-2 position-relative miniIcon"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 48C141.13 48 48 141.13 48 256s93.13 208 208 208 208-93.13 208-208S370.87 48 256 48z"
                    fill="none"
                    stroke-miterlimit="10"
                    stroke-width="32"
                  />
                  <path
                    d="M256 48c-58.07 0-112.67 93.13-112.67 208S197.93 464 256 464s112.67-93.13 112.67-208S314.07 48 256 48z"
                    fill="none"
                    stroke-miterlimit="10"
                    stroke-width="32"
                  />
                  <path
                    d="M117.33 117.33c38.24 27.15 86.38 43.34 138.67 43.34s100.43-16.19 138.67-43.34m0 277.34c-38.24-27.15-86.38-43.34-138.67-43.34s-100.43 16.19-138.67 43.34"
                    fill="none"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="32"
                  />
                  <path
                    d="M256 48v416m208-208H48"
                    fill="none"
                    stroke-miterlimit="10"
                    stroke-width="32"
                  />
                </svg>
                <div
                  class="truncateWrapper"
                >
                  <a
                    aria-label="Website link"
                    class="link d-block text-truncate text-dark link"
                    href="https://cloudevents.io"
                    rel="noopener noreferrer"
                    tabindex="-1"
                    target="_blank"
                  >
                    <div
                      class="d-flex flex-row align-items-baseline"
                    >
                      <div
                        class="text-truncate"
                      >
                        https://cloudevents.io
                      </div>
                      <svg
                        class="ms-2 icon"
                        fill="none"
                        height="1em"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
                        />
                        <polyline
                          points="15 3 21 3 21 9"
                        />
                        <line
                          x1="10"
                          x2="21"
                          y1="14"
                          y2="3"
                        />
                      </svg>
                    </div>
                  </a>
                </div>
              </div>
              <div
                class="d-flex flex-row align-items-center my-1 link"
              >
                <svg
                  class="me-2 position-relative miniIcon"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 48C141.13 48 48 141.13 48 256s93.13 208 208 208 208-93.13 208-208S370.87 48 256 48z"
                    fill="none"
                    stroke-miterlimit="10"
                    stroke-width="32"
                  />
                  <path
                    d="M256 48c-58.07 0-112.67 93.13-112.67 208S197.93 464 256 464s112.67-93.13 112.67-208S314.07 48 256 48z"
                    fill="none"
                    stroke-miterlimit="10"
                    stroke-width="32"
                  />
                  <path
                    d="M117.33 117.33c38.24 27.15 86.38 43.34 138.67 43.34s100.43-16.19 138.67-43.34m0 277.34c-38.24-27.15-86.38-43.34-138.67-43.34s-100.43 16.19-138.67 43.34"
                    fill="none"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="32"
                  />
                  <path
                    d="M256 48v416m208-208H48"
                    fill="none"
                    stroke-miterlimit="10"
                    stroke-width="32"
                  />
                </svg>
                <div
                  class="truncateWrapper"
                >
                  <a
                    aria-label="Website link"
                    class="link d-block text-truncate text-dark link"
                    href="https://cloudevents.io/test"
                    rel="noopener noreferrer"
                    tabindex="-1"
                    target="_blank"
                  >
                    <div
                      class="d-flex flex-row align-items-baseline"
                    >
                      <div
                        class="text-truncate"
                      >
                        https://cloudevents.io/test
                      </div>
                      <svg
                        class="ms-2 icon"
                        fill="none"
                        height="1em"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
                        />
                        <polyline
                          points="15 3 21 3 21 9"
                        />
                        <line
                          x1="10"
                          x2="21"
                          y1="14"
                          y2="3"
                        />
                      </svg>
                    </div>
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-expanded="false"
        class="cursorDefault"
      >
        <div
          class="d-flex flex-row align-items-center link"
        >
          <svg
            class="icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 48C141.13 48 48 141.13 48 256s93.13 208 208 208 208-93.13 208-208S370.87 48 256 48z"
              fill="none"
              stroke-miterlimit="10"
              stroke-width="32"
            />
            <path
              d="M256 48c-58.07 0-112.67 93.13-112.67 208S197.93 464 256 464s112.67-93.13 112.67-208S314.07 48 256 48z"
              fill="none"
              stroke-miterlimit="10"
              stroke-width="32"
            />
            <path
              d="M117.33 117.33c38.24 27.15 86.38 43.34 138.67 43.34s100.43-16.19 138.67-43.34m0 277.34c-38.24-27.15-86.38-43.34-138.67-43.34s-100.43 16.19-138.67 43.34"
              fill="none"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="32"
            />
            <path
              d="M256 48v416m208-208H48"
              fill="none"
              stroke-miterlimit="10"
              stroke-width="32"
            />
          </svg>
          <div
            class="ms-1"
          >
            Websites
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
